subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultArtifactTransformDependenciesProvider.java

```diff
@@ -100,6 +104,18 @@ public ArtifactTransformDependenciesInternal forAttributes(ImmutableAttributes a
         return new DefaultArtifactTransformDependencies(files);
     }
 
+    private Set<ComponentIdentifier> initializeDependencyIdentifiers() {
+        ResolutionResult resolutionResult = resolvableDependencies.getResolutionResult();
+        Set<ComponentIdentifier> result = Sets.newHashSet();
```

Comment:
Good point, will update

Reply:
I can make the field an immutable set but I cannot use the builder, as I rely on the `Set.add` result to prevent looping when there are cycles in the dependencies.